### PR TITLE
[2.11] Implement inline errors to the Add Backend form

### PR DESCRIPTION
**Pull request by josemigallas**
**What this PR does / why we need it**:

Renders the following inline errors whenever the form is sent with invalid data:
- Backend not found: this should never happen but the submit button can be disabled manually and the form submitted (see image).
Screenshot 2021-04-15 at 18 28 51

- Path already taken: this will be returned by the server. In this case, the selected backend will still be selected despite the page refresh.

Demo:
add-backend-usage

**Which issue(s) this PR fixes** 

[THREESCALE-6883: Implements inline validations and flash errors upon submition](https://issues.redhat.com/browse/THREESCALE-6883)

> Subtask of [3scale][2.11][HI-prio] Improve Select in 'Add Backend to Product' form
[THREESCALE-6882: Embed New Backend form into a modal](https://issues.redhat.com/browse/THREESCALE-6882)

### CODE REVIEW

**Review by thomasmaas — DISMISSED** [review]

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -236,8 +236,10 @@
       location.href = "?view=" + view + "&s=" + param + "&for_category=" + category;
     });
 
+    // DEPRECATED: since the introduction of PF4 and React, colorbox is being removed. Also jquery-ujs has been replaced with rails-ujs.
+    // Added #colorbox selector to target only non-React forms
     // show errors from ajax in formtastic
-    $('form').live('ajax:error', function(event, xhr, status, error) {
+    $('#colorbox form').live('ajax:error', function(event, xhr, status, error) {
```
Comment (damianpm) [inline app/assets/javascripts/application.js:242]: colorbox is also used in the Developer Portal, have you checked it doesn't affect old dev portals?
(I don't remember if this fie is loaded in dev portal)

File: app/javascript/src/BackendApis/components/NewBackendModal.jsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// @flow
+
+import * as React from 'react'
+
```
Comment (damianpm) [inline app/javascript/src/BackendApis/components/NewBackendModal.jsx:4]: For consistency with `NewBackendForm.jsx`, you coud import `useState` and `useEffect`

File: app/javascript/src/BackendApis/components/NewBackendModal.jsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      const backend = JSON.parse(xhr.responseText)
+      onCreateBackend(backend)
+    } else if (status === 'error') {
+      const errors = JSON.parse(xhr.responseText)
+      setErrors(errors)
+    }
+  }
+
+  React.useEffect(() => {
+    $('form#new_backend_api_config')
```
Comment (damianpm) [inline app/javascript/src/BackendApis/components/NewBackendModal.jsx:43]: Why we need jQuery here, can we add listeners in other way?
If complicated and out of scope, we could add a todo

Comment (josemigallas) [reply app/javascript/src/BackendApis/components/NewBackendModal.jsx]: This has to do with the way Rails makes async calls. I think we haven't done this with React before but the library is injected into jQuery so I'm afraid this is how it's done.. ugly yes. We can't even import $ from node_modules because that's a different version and it won't include jquery-ujs.

In Rails 5 there's a new module that replaces jquery-ujs, however porta uses it extensively and migrating to rails-ujs (the new module) would be a massive task. There are already some todos pointing to this. But again, I think it's not realistic to think we're going to "fix" this anytime soon. At least we'd need to migrate everything to React and stop using colorbox and the rest of old libraries.

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (josemigallas) [reply app/assets/javascripts/application.js]: I'll take a more thorough look. What I'm trying to do is to prevent jquery-ujs from injecting callbacks in our new components. Maybe it's better to do it the other way around: add an excluding selector to filter our PF forms `$('form:not(.pf-c-form)')`

File: app/javascript/src/BackendApis/components/NewBackendModal.jsx (hunk 3 above)
Comment (josemigallas) [reply app/javascript/src/BackendApis/components/NewBackendModal.jsx]: Actually I just saw there's a `$FlowFixMe` in the next line. These are to be treated as TODO.

File: app/lib/three_scale/oauth2/github_client.rb [hunk 4]
```diff
@@ -49,7 +49,6 @@ def primary_email
       end
 
       def emails
-        access_token.options[:mode] = :query
```
Comment (hallelujah) [inline app/lib/three_scale/oauth2/github_client.rb:52]: This is not related to this PR I think, probably need to rebase on master

**Review by hallelujah — DISMISSED** [review]

File: app/lib/three_scale/oauth2/github_client.rb [hunk 5]
```diff
@@ -49,7 +49,6 @@ def primary_email
       end
 
       def emails
```
Comment (josemigallas) [inline app/lib/three_scale/oauth2/github_client.rb:51]:
```suggestion
      def emails
        access_token.options[:mode] = :query
```

File: Gemfile.prod [hunk 6]
```diff
@@ -1,8 +1 @@
 eval_gemfile 'Gemfile.base'
-
```
Comment (damianpm) [inline Gemfile.prod:2]: I think you have to rebase, see https://github.com/3scale/porta/pull/2423

**Review by damianpm — CHANGES_REQUESTED** [review]

**Review by damianpm — APPROVED** [review]

**Review by hallelujah — APPROVED** [review]
